File: bindings/java/src/main/java/io/angzarr/router/ProjectorDispatch.java
package io.angzarr.router;

import com.google.protobuf.Message;
import io.angzarr.router.Thunks.ProjectorEventThunk;
import io.angzarr.router.Thunks.ProjectorFinishThunk;
import io.angzarr.router.Thunks.ProjectorUnknownThunk;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

/**
 * One projector component's registration: its name, projection-state factory,
 * the domains it folds, per-event fold thunks, and the finisher that carries the
 * cover onto the Projection.
 */
public final class ProjectorDispatch {
  final String name;
  final Supplier<Message.Builder> factory;
  List<String> domains = List.of();
  final Map<String, ProjectorEventThunk> events = new LinkedHashMap<>();
  ProjectorFinishThunk finish;
  ProjectorUnknownThunk unknown;

  public ProjectorDispatch(String name, Supplier<Message.Builder> factory) {
    this.name = name;
    this.factory = factory;
  }

  /** Declares the domains this projector folds. */
  public ProjectorDispatch forDomains(String... domains) {
    this.domains = List.of(domains);
    return this;
  }

  /** Registers the fold thunk for a fully-qualified event type. */
  public ProjectorDispatch onEvent(String fullName, ProjectorEventThunk thunk) {
    events.put(fullName, thunk);
    return this;
  }

  /** Registers the finisher that produces the Projection from the folded state. */
  public ProjectorDispatch finish(ProjectorFinishThunk thunk) {
    this.finish = thunk;
    return this;
  }

  /** Registers an optional observer for events outside the declared set. */
  public ProjectorDispatch onUnknown(ProjectorUnknownThunk thunk) {
    this.unknown = thunk;
    return this;
  }
}

File: bindings/csharp/src/ProjectorDispatch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Google.Protobuf;

namespace Angzarr.Router;

/// <summary>
/// One projector component's registration: its name, projection-state factory,
/// the domains it folds, per-event fold thunks, and the finisher that carries
/// the cover onto the Projection. Generic in the projection state message.
/// </summary>
public sealed class ProjectorDispatch<TState>
    where TState : class, IMessage
{
    internal readonly string Name;
    internal readonly Func<TState> Factory;
    internal IReadOnlyList<string> Domains = Array.Empty<string>();
    internal readonly Dictionary<string, ProjectorEventThunk<TState>> Events = new();
    internal ProjectorFinishThunk<TState>? FinishThunk;
    internal ProjectorUnknownThunk? Unknown;

    public ProjectorDispatch(string name, Func<TState> factory)
    {
        Name = name;
        Factory = factory;
    }

    /// <summary>Declares the domains this projector folds.</summary>
    public ProjectorDispatch<TState> ForDomains(params string[] domains)
    {
        Domains = domains.ToList();
        return this;
    }

    /// <summary>Registers the fold thunk for a fully-qualified event type.</summary>
    public ProjectorDispatch<TState> OnEvent(string fullName, ProjectorEventThunk<TState> thunk)
    {
        Events[fullName] = thunk;
        return this;
    }

    /// <summary>Registers the finisher that produces the Projection from the
    /// folded state.</summary>
    public ProjectorDispatch<TState> Finish(ProjectorFinishThunk<TState> thunk)
    {
        FinishThunk = thunk;
        return this;
    }

    /// <summary>Registers an optional observer for events outside the declared
    /// set.</summary>
    public ProjectorDispatch<TState> OnUnknown(ProjectorUnknownThunk thunk)
    {
        Unknown = thunk;
        return this;
    }
}
